File: MobileAppReactNativeMobX/src/business/state/Portfolio.ts
import { makeAutoObservable, reaction } from 'mobx';
import AvailableStock from './AvailableStock';
import CashBalance from './CashBalance';
import Stock from './Stock';
import { StorageManager } from '../dao/StorageManager';
import { round } from '../utils/utils';
import { BuyOrSell } from './BuyOrSell';

class Portfolio {
  stocks: Stock[];
  cashBalance: CashBalance;

  constructor(stocks?: Stock[], cashBalance?: CashBalance) {
    this.stocks = stocks ?? [];
    this.cashBalance = cashBalance ?? new CashBalance(0);
    makeAutoObservable(this);

    // When the Portfolio changes, mark it for later saving to the local disk.
    reaction(
      () => [
        this.stocks.map(stock => stock),
        this.cashBalance,
        this.cashBalance.amount],
      () => StorageManager.markPortfolioChanged(),
      {
        // Prevents the reaction from running immediately.
        fireImmediately: false,
      },
    );
  }

  get isEmpty(): boolean {
    return this.stocks.length === 0;
  }

  clearStock(ticker: string) {
    this.setStockInPortfolio(ticker, 0, 0);
  }

  /**
   * Adds {@link howMany} shares of the given {@link availableStock}.
   * Will not chance the cash balance.
   */
  addStock(availableStock: AvailableStock, howMany: number) {

    let pos = this.getStockPositionInList(availableStock);

    if (pos === -1) {
      this.stocks.push(availableStock.toStock(howMany));
    } else {
      let stock = this.stocks[pos];

      const newShares = stock.howManyShares + howMany;

      stock.averagePrice = round(
        ((stock.howManyShares * stock.averagePrice) +
          (howMany * availableStock.currentPrice)) /
        newShares);

      stock.howManyShares = newShares;
    }
  }

  clearAll() {
    this.stocks.length = 0;
  }

  setStockInPortfolio(ticker: string,
                      quantity: number,
                      averagePrice: number) {

    this.stocks = this.stocks.filter(stock => stock.ticker !== ticker);

    if (quantity > 0) {
      const newStock = new Stock(ticker, quantity, averagePrice);
      this.stocks.push(newStock);
    }
  }

  howManyStocks(ticker: string): number {
    const stock = this.getStockOrNull(ticker);
    return stock ? stock.howManyShares : 0;
  }

  getStock(ticker: string): Stock {
    let stock = this.getStockOrNull(ticker);
    if (!stock) throw new Error(`Stock '${ticker}' not found.`);
    return stock;
  }

  getStockOrNull(ticker: string): Stock | null {
    return this.stocks.find(stock => stock.ticker === ticker) ?? null;
  }

  hasStock(availableStock: AvailableStock): boolean {
    return this.getStockPositionInList(availableStock) !== -1;
  }

  /** Returns true if the portfolio contains the given stock. */
  hasMoneyToBuyStock(availableStock: AvailableStock): boolean {
    return this.cashBalance.amount >= availableStock.currentPrice;
  }

  /** Will {@link buyOrSell} the amount of {@link howMany} shares of the given {@link availableStock}. */
  buyOrSell(buyOrSell: BuyOrSell,
            availableStock: AvailableStock,
            howMany: number,
  ) {
    if (buyOrSell === BuyOrSell.BUY) {
      this.buy(availableStock, howMany);
    } else {
      this.sell(availableStock, howMany);
    }
  }

  /**
   * Buy {@link howMany} shares of the given {@link availableStock}].
   * Will reduce the cash balance accordingly.
   * Throws an exception if there is not enough money.
   */
  buy(availableStock: AvailableStock, howMany: number) {

    if (this.cashBalance.amount < availableStock.currentPrice * howMany) {
      throw new Error('Not enough money to buy stock');
    }
    //
    else {
      this.cashBalance.remove(availableStock.currentPrice * howMany);
      this.addStock(availableStock, howMany);
    }
  }

  /**
   * Sell {@link howMany} shares of the given {@link availableStock}.
   * Will increase the cash balance accordingly.
   * Throws an exception if you do not own the stock.
   */
  sell(availableStock: AvailableStock, howMany: number) {
    const pos = this.getStockPositionInList(availableStock);

    if (pos === -1) {
      throw new Error('Cannot sell stock you do not own');
    }
    //
    else {
      let stock = this.stocks[pos];

      if (stock.howManyShares < howMany) {
        throw new Error(`Cannot sell ${howMany} shares of stock you do not own`);
      }
      // Remove the stock entirely if all shares are sold.
      else if (stock.howManyShares === howMany) {
        this.stocks.splice(pos, 1);
      }
      //
      else {
        const newShares = stock.howManyShares - howMany;

        // Recalculate the averagePrice. As mentioned, this would be unusual in a real-world scenario.
        stock.averagePrice = round(
          ((stock.howManyShares * stock.averagePrice) -
            (howMany * availableStock.currentPrice)) /
          newShares);

        stock.howManyShares = newShares;
      }

      // Increase the cash balance.
      this.cashBalance.add(availableStock.currentPrice * howMany);
    }
  }

  private getStockPositionInList(availableStock: AvailableStock): number {
    return this.stocks.findIndex(stock => stock.ticker === availableStock.ticker);
  }

  get totalCostBasis(): number {
    return this.stocks.reduce((sum, stock) => sum + stock.costBasis, 0)
      + this.cashBalance.amount;
  }

  copyFrom(other: Portfolio) {
    this.stocks = other.stocks.map(stock => stock);
    this.cashBalance = other.cashBalance;
  }

  toString(): string {
    return `                   
        stocks: ${this.stocks}    
        cashBalance: ${this.cashBalance}    
        `;
  }
}

export default Portfolio;

File: MobileAppFlutterCelest/celest/lib/my_src/models/portfolio.dart
import 'package:async_redux_core/async_redux_core.dart';
import 'package:celest_backend/exceptions.dart';
import 'package:celest_backend/models.dart';
import 'package:celest_backend/my_src/models/buy_or_sell.dart';
import 'package:celest_backend/my_src/models/cash_balance.dart';
import 'package:celest_backend/my_src/models/utils/map_deserialization_extension.dart';
import 'package:collection/collection.dart';
import 'package:fast_immutable_collections/fast_immutable_collections.dart';
import 'package:i18n_extension_core/default.i18n.dart';
import 'package:meta/meta.dart';

import 'utils/json.dart';

@immutable
class Portfolio {
  static const Portfolio EMPTY = Portfolio._();

  final IList<Stock> stocks;
  final CashBalance cashBalance;

  Portfolio({
    Iterable<Stock>? stocks,
    this.cashBalance = CashBalance.ZERO,
  }) : stocks = IList.orNull(stocks) ?? const IListConst([]);

  const Portfolio._()
      : stocks = const IListConst([]),
        cashBalance = CashBalance.ZERO;

  Portfolio copyWith({
    Iterable<Stock>? stocks,
    CashBalance? cashBalance,
  }) {
    return Portfolio(
      stocks: (stocks != null) ? stocks.toIList() : this.stocks,
      cashBalance: cashBalance ?? this.cashBalance,
    );
  }

  bool get isEmpty => stocks.isEmpty;

  Portfolio withCashBalance(CashBalance newCashBalance) => copyWith(cashBalance: newCashBalance);

  Portfolio addCashBalance(double howMuch) => copyWith(cashBalance: cashBalance.add(howMuch));

  Portfolio removeCashBalance(double howMuch) {
    if (cashBalance.amount < howMuch) {
      throw UserException('Not enough money to remove.'.i18n);
    }

    return copyWith(cashBalance: cashBalance.remove(howMuch));
  }

  Portfolio withoutStock(String ticker) => withStock(ticker, 0, 0.0);

  Portfolio withAddedStock(AvailableStock availableStock, int howMany) {
    final pos = _getStockPositionInList(availableStock);
    List<Stock> newStocks;

    if (pos == -1) {
      newStocks = [...stocks, availableStock.toStock(shares: howMany)];
    } else {
      final stock = stocks[pos];
      final newShares = stock.howManyShares + howMany;
      final newAveragePrice = round(
          ((stock.howManyShares * stock.averagePrice) + (howMany * availableStock.currentPrice)) /
              newShares);

      newStocks = [...stocks];
      newStocks[pos] = Stock(
        stock.ticker,
        howManyShares: newShares,
        averagePrice: newAveragePrice,
      );
    }

    return copyWith(stocks: newStocks);
  }

  Portfolio withoutStocks() {
    return copyWith(stocks: []);
  }

  Portfolio withStock(String ticker, int quantity, double averagePrice) {
    final newStocks = stocks.where((stock) => stock.ticker != ticker).toList();

    if (quantity > 0) {
      final newStock = Stock(ticker, howManyShares: quantity, averagePrice: averagePrice);
      newStocks.add(newStock);
    }

    return copyWith(stocks: newStocks);
  }

  int howManyStocks(String ticker) {
    final stock = getStockOrNull(ticker);
    return stock?.howManyShares ?? 0;
  }

  Stock getStock(String ticker) {
    final stock = getStockOrNull(ticker);
    if (stock == null) {
      throw UserException('Stock %s not found.'.i18n.fill([ticker]));
    }
    return stock;
  }

  Stock? getStockOrNull(String ticker) {
    return stocks.firstWhereOrNull((stock) => stock.ticker == ticker);
  }

  bool hasStock(AvailableStock availableStock) {
    return _getStockPositionInList(availableStock) != -1;
  }

  bool hasMoneyToBuyStock(AvailableStock availableStock) {
    return cashBalance.amount >= availableStock.currentPrice;
  }

  Portfolio buyOrSell(BuyOrSell buyOrSell, AvailableStock availableStock, int howMany) {
    return (buyOrSell == BuyOrSell.buy)
        ? buy(availableStock, howMany: howMany)
        : sell(availableStock, howMany: howMany);
  }

  /// Buys [howMany] shares of [availableStock].
  /// If the user does not have enough money, throws a [UserException].
  Portfolio buy(AvailableStock availableStock, {required int howMany}) {
    if (cashBalance.amount < availableStock.currentPrice * howMany) {
      throw UserException('Not enough money to buy stock'.i18n);
    } else {
      final newCashBalance =
          CashBalance(cashBalance.amount - availableStock.currentPrice * howMany);
      final newPortfolio = withAddedStock(availableStock, howMany);
      return newPortfolio.copyWith(cashBalance: newCashBalance);
    }
  }

  /// Sells [howMany] shares of [availableStock].
  /// If the user does not own the stock, or does not own enough shares, throws a [UserException].
  /// This exception will be shown to the user in a dialog.
  Portfolio sell(AvailableStock availableStock, {required int howMany}) {
    final pos = _getStockPositionInList(availableStock);

    if (pos == -1) {
      throw UserException('Cannot sell stock you do not own'.i18n);
    } else {
      final stock = stocks[pos];

      if (stock.howManyShares < howMany) {
        throw UserException('Cannot sell %d shares of stock you do not own'.i18n.fill([howMany]));
      } else {
        final newShares = stock.howManyShares - howMany;
        final newAveragePrice = round(
            ((stock.howManyShares * stock.averagePrice) - (howMany * availableStock.currentPrice)) /
                newShares);

        var newStocks = [...stocks];
        newStocks[pos] = Stock(
          stock.ticker,
          howManyShares: newShares,
          averagePrice: newAveragePrice,
        );

        if (newShares == 0) {
          newStocks = newStocks.where((stock) => stock.ticker != availableStock.ticker).toList();
        }

        final newCashBalance =
            CashBalance(cashBalance.amount + availableStock.currentPrice * howMany);
        return copyWith(stocks: newStocks, cashBalance: newCashBalance);
      }
    }
  }

  int _getStockPositionInList(AvailableStock availableStock) {
    return stocks.indexWhere((stock) => stock.ticker == availableStock.ticker);
  }

  double get totalCostBasis {
    return stocks.fold(0.0, (sum, stock) => sum + stock.costBasis) + cashBalance.amount;
  }

  @override
  String toString() => 'Portfolio{stocks: $stocks, cashBalance: $cashBalance}';

  Map<String, dynamic> toJson() => {
        'stocks': stocks.map((stock) => stock.toJsonPersistor()).toList(),
        'cashBalance': cashBalance.toJsonPersistor(),
      };

  static Portfolio fromJson(Json? json) {
    if (json == null)
      return Portfolio.EMPTY;
    else {
      IList<Stock> stocks = json.asIListOf('stocks', Stock.fromJsonPersistor);
      CashBalance cashBalance = json.asCashBalance('cashBalance') ?? CashBalance.ZERO;
      return Portfolio(stocks: stocks, cashBalance: cashBalance);
    }
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Portfolio &&
          runtimeType == other.runtimeType &&
          stocks == other.stocks &&
          cashBalance == other.cashBalance;

  @override
  int get hashCode => stocks.hashCode ^ cashBalance.hashCode;
}
